### PR TITLE
Detecting OLED panels from EDID data

### PR DESCRIPTION
**Pull request by Herbstein**
This is related to pop-os/system76-oled#5.

I've made these changes with two patches applied. The polkit downgrade and the OLED detection patches. I figured it would be easier to talk about changes if the branch compiles as-is.

### CODE REVIEW

**Comment by Herbstein** [comment]
I just want to point out that the vendor code used in the `known_oled` array is generated with a function similar to the one below. The input in the case of the panel in the XPS7590 and Adder WS would be an array containing `S`, `D`, and `C`.

```c
unsigned short convert_vendor(char vendor[3]) {
    unsigned short mask = 0x1F;
    unsigned short i0 = 'A' - 1;
    unsigned short v1 = ((vendor[0] - i0) & mask) << 10;
    unsigned short v2 = ((vendor[1] - i0) & mask) <<  5;
    unsigned short v3 = ((vendor[2] - i0) & mask) <<  0;
    return v1 | v2 | v3;
}
```

Of course, if you have access to the EDID information reading it straight from there is easier but I wanted to post this for posterity.

**Comment by Herbstein** [comment]
Theres 4 `TODO` comments in the code with questions/deliberations. My first iterations did direct `memcmp` calls, but it has since been refactored to `usigned long` and `unsigned short` comparisons. The optimized assembly generally comes out the same, and (to me) the integer comparisons are easier to follow.

I haven't been able to test this properly, but [I inserted the first 12 bytes of my own EDID file](https://godbolt.org/z/S4iAn_) (from an OLED xps7590), and that does indeed result in a `TRUE` return value.

**Comment by jackpot51** [comment]
Thanks, I will take a look soon

**Comment by Herbstein** [comment]
Any update on this?

File: meson.build [hunk 1]
```diff
@@ -98,7 +98,7 @@ libgeoclue_dep = dependency('libgeoclue-2.0', version: '>= 2.3.1')
 libnotify_dep = dependency('libnotify', version: '>= 0.7.3')
 libpulse_mainloop_glib_dep = dependency('libpulse-mainloop-glib', version: '>= 2.0')
 pango_dep = dependency('pango', version: '>= 1.20.0')
-polkit_gobject_dep = dependency('polkit-gobject-1', version: '>= 0.114')
+polkit_gobject_dep = dependency('polkit-gobject-1', version: '>= 0.105')
```
Comment (jackpot51) [inline meson.build]: What is the purpose of this change? Could you undo it?

File: plugins/color/gsd-color-state.c [hunk 2]
```diff
@@ -46,6 +47,70 @@
...
+  { 19587, 41001 }
+};
+
+static gboolean
+has_oled_builtin_display (void)
+{
+  gboolean has_oled = FALSE;
+
+  // TODO: The Rust code does a dir walk to find all outputs. Needed?
+  FILE *f = fopen("/sys/class/drm/card0-eDP-1/edid", "rb");
```
Comment (jackpot51) [inline plugins/color/gsd-color-state.c]: This is okay for now, if both the supported models end up having the OLED output at the same path. But I would be concerned if the detection orders of GPUs is not deterministic.

File: plugins/color/gsd-color-state.c [hunk 3]
```diff
@@ -46,6 +47,70 @@
...
+    memcpy(&header.info.vendor, edid + 8, 2);
+    memcpy(&header.info.product, edid + 10, 2);
+
+    // TODO: Endianess swap. Swap here or change the value in the array?
+    unsigned short vendor = header.info.vendor;
+    header.info.vendor = (vendor >> 8) | (vendor << 8);
+    
+    if (header.pattern == 0xFFFFFFFFFFFF00) {
+      // TODO: I don't like this line. Alternatives?
+      unsigned long known_oled_len = sizeof(known_oled) / sizeof(known_oled[0]);
```
Comment (jackpot51) [inline plugins/color/gsd-color-state.c]: This is a pretty standard way to get the size of a statically defined array in C, you can remove the TODO

File: plugins/color/gsd-color-state.c [hunk 4]
```diff
@@ -20,6 +20,7 @@
 
 #include "config.h"
 
+#include <ctype.h>
```
Comment (jackpot51) [inline plugins/color/gsd-color-state.c]: You made edits directly to the file instead of to the patch. If you can, use quilt and move this into the patch. If you cannot, let me know and I can take care of it later

File: plugins/color/gsd-color-state.c (hunk 2 above)
Comment (Herbstein) [reply plugins/color/gsd-color-state.c]: Would the long-term solution be to interact with the EDID data being passed around elsewhere in the file?

File: plugins/color/gsd-color-state.c (hunk 3 above)
Comment (Herbstein) [reply plugins/color/gsd-color-state.c]: Thanks! I still have a hard time distinguishing kludges from standard C.

File: plugins/color/gsd-color-state.c (hunk 4 above)
Comment (Herbstein) [reply plugins/color/gsd-color-state.c]: The changes should be moved to the patch file now.

**Comment by Herbstein** [comment]
I wasn't aware of quilt, but you mentioning it lead me to a very helpful Debian site that made it all very easy. Thanks!

**Review by jackpot51 — APPROVED** [review]

**Comment by jackpot51** [comment]
It looks good now. 20.04 release is right around the corner so it would be good to put this there instead by updating the patch in master_focal

**Comment by jackpot51** [comment]
@Herbstein I just force pushed this, rebasing it on the master_focal branch. Will be merging it after it builds and I can verify it works on the System76 Adder WS

**Comment by Herbstein** [comment]
Sounds great! Looking forward to having this fixed.

**Review by leviport — APPROVED** [review]
Works on addw1!

**Review by jackpot51 — APPROVED** [review]

**Comment by jackpot51** [comment]
Works on addw2 as well

**Comment by michalmielczynski** [comment]
Works on my XPS 15. Great works guys, THANK YOU so much!
